### PR TITLE
statistics: remove rlock using copy on write.

### PR DESCRIPTION
**Pull request by hanfei1991**
every time we update stats cache, we copy the whole map. By this way we can avoid rlock.
@shenli @coocood @zimulala @lamxTyler PTAL

### CODE REVIEW

File: statistics/statscache.go [hunk 1]
```diff
@@ -72,47 +77,51 @@ func (h *Handle) Update(is infoschema.InfoSchema) error {
...
+	for k, v := range oldCache {
+		newCache[k] = v
+	}
+	return newCache
+}
+
+// UpdateTableStats updates the statistics table cache using copy on write.
+func (h *Handle) UpdateTableStats(stats []*statsInfo) {
 	h.m.Lock()
 	defer h.m.Unlock()
```
Comment (coocood) [inline statistics/statscache.go]: The lock can be removed?

**Comment by zimulala** [comment]
LGTM

File: statistics/statscache.go [hunk 2]
```diff
@@ -25,31 +25,28 @@ import (
...
 // Handle can update stats info periodically.
 type Handle struct {
 	ctx         context.Context
 	lastVersion uint64
-	cache       map[int64]*statsInfo
-	m           sync.RWMutex
+	cache       atomic.Value
 }
 
 // Clear the statsTblCache, only for test.
```
Comment (lamxTyler) [inline statistics/statscache.go]: It's not `statsTblCache`?

File: statistics/statscache.go [hunk 3]
```diff
@@ -72,47 +70,45 @@ func (h *Handle) Update(is infoschema.InfoSchema) error {
 		// statistics can not be used any more, we give it a pseudo one.
```
Comment (lamxTyler) [inline statistics/statscache.go]: Update the comment?

File: executor/analyze.go [hunk 4]
```diff
@@ -115,8 +116,18 @@ func (e *AnalyzeExec) buildStatisticsAndSaveToKV(count int64, columnSamples [][]
...
+	dom := sessionctx.GetDomain(e.ctx)
+	err = dom.StatsHandle().SaveToStorage(e.ctx, t)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	lease := dom.DDL().GetLease()
+	if lease > 0 {
+		time.Sleep(lease * 2)
+	} else {
+		dom.StatsHandle().Update(GetInfoSchema(e.ctx))
```
Comment (zimulala) [inline executor/analyze.go]: Check the error.

**Comment by lamxTyler** [comment]
LGTM

File: statistics/statscache.go [hunk 5]
```diff
@@ -68,51 +66,48 @@ func (h *Handle) Update(is infoschema.InfoSchema) error {
...
-	defer h.m.Unlock()
-	stats, ok := h.cache[id]
-	if !ok {
-		si := &statsInfo{
-			tbl:     statsTbl,
-			version: version,
-		}
-		h.cache[id] = si
-		return
+func (h *Handle) loadFromOldCache() statsCache {
```
Comment (coocood) [inline statistics/statscache.go]: This name doesn't indicate the returned cache is new cache.

File: executor/analyze.go [hunk 6]
```diff
@@ -115,8 +116,21 @@ func (e *AnalyzeExec) buildStatisticsAndSaveToKV(count int64, columnSamples [][]
...
-	err = sessionctx.GetDomain(e.ctx).StatsHandle().SaveToStorage(e.ctx, t)
-	return errors.Trace(err)
+	dom := sessionctx.GetDomain(e.ctx)
+	err = dom.StatsHandle().SaveToStorage(e.ctx, t)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	lease := dom.DDL().GetLease()
+	if lease > 0 {
+		time.Sleep(lease * 2)
```
Comment (coocood) [inline executor/analyze.go]: Add comment about the sleep.

**Pull request merged**